jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java

```diff
@@ -2585,19 +2584,18 @@ public ClassLoader getClassLoader()
 
             // no security manager just return the classloader
             if (!_usingSecurityManager)
+            {
                 return _classLoader;
+            }
             else
             {
                 // check to see if the classloader of the caller is the same as the context
-                // classloader, or a parent of it
+                // classloader, or a parent of it, as required by the javadoc specification.
                 try
                 {
-                    Class<?> reflect = Loader.loadClass("sun.reflect.Reflection");
-                    Method getCallerClass = reflect.getMethod("getCallerClass",Integer.TYPE);
-                    Class<?> caller = (Class<?>)getCallerClass.invoke(null,2);
-
+                    Class<?> caller = new Caller().getCallerClass(2);
                     boolean ok = false;
-                    ClassLoader callerLoader = caller.getClassLoader();
+                    ClassLoader callerLoader = caller == null ? null : caller.getClassLoader();
```

Comment:
Couldn't we skip this null check by having `new Caller().getCallerClassLoader(2)` instead?

Reply:
I don't understand how? If depth 2 is not valid, what `ClassLoader` do I return?

Reply:
What i'm proposing ...

From:
``` java
Class<?> caller = new Caller().getCallerClass(2);
boolean ok = false;
ClassLoader callerLoader = caller == null ? null : caller.getClassLoader();
...
    private static class Caller extends SecurityManager
    {
        public Class<?> getCallerClass(int depth)
        {
            if (depth < 0)
                return null;
            Class<?>[] classContext = getClassContext();
            if (classContext.length <= depth)
                return null;
            return classContext[depth];
        }
    }
```

To:
``` java
boolean ok = false;
ClassLoader callerLoader = new Caller().getCallerClassLoader(2);
...
    private static class Caller extends SecurityManager
    {
        public ClassLoader<?> getCallerClassLoader(int depth)
        {
            if (depth < 0)
                return null;
            Class<?>[] classContext = getClassContext();
            if (classContext.length <= depth)
                return null;
            return classContext[depth].getClassLoader();
        }
    }
```

You are already doing a better job at validity checking down in the `Caller`.
Plus if you get past your 2 range checks, is it possible to have a null classContext entry??